### PR TITLE
Fix Linux builds

### PR DESCRIPTION
**Pull request by balintfodor**
This is my first pull request to bincrafters so don't be too hard on me if something is off :) I'm here to learn.

This PR fixes building for GCC 8, 9 and Clang 6-9, so it's not all that got green but I'm happy with it so far.

I'm not sure if it's ok to open the pull request against `testing/20200508` or it should go to a new branch like `testing/2021XXXX`?

I'm not sure about the bincrafters release workflow either, when and how should it get to a `stable/2021XXXX` branch and how will the prebuilt package get to the artifact repo?

### CODE REVIEW

**Comment by balintfodor** [comment]
This is weird, the same thing passed on appveyor and GHA/Windows* for my branch https://github.com/balintfodor/conan-crashpad/actions/runs/644114887 but looks like this PR's checker failed on them.

File: conanfile.py [hunk 1]
```diff
@@ -99,6 +99,12 @@ def _setup_args_gn(self):
         self._set_env_arg(args, "CXXFLAGS", "extra_cflags_objcc")
         self._set_env_arg(args, "LDFLAGS",  "extra_ldflags")
 
+        cxx = os.getenv("CXX")
+        if cxx and "g++" in cxx:
```
Comment (ericLemanissier) [inline conanfile.py]:
```suggestion
        if self.settings.compiler == "gcc":
```
Does it work like this ?

**Comment by ericLemanissier** [comment]
> This is my first pull request to bincrafters so don't be too hard on me if something is off :) I'm here to learn.

Thanks for the PR !

> I'm not sure if it's ok to open the pull request against `testing/20200508` or it should go to a new branch like `testing/2021XXXX`?

This is the correct branch for a pull request

> I'm not sure about the bincrafters release workflow either, when and how should it get to a `stable/2021XXXX` branch and how will the prebuilt package get to the artifact repo?

If most of the CI is green, and other contributors are ok with the state of the recipe, we will create `stable/20200508` branch on top of `testing/2020.5.8`

File: test_package/conanfile.py [hunk 2]
```diff
@@ -7,6 +7,8 @@ class TestPackageConan(ConanFile):
     generators = "cmake"
 
     def build(self):
+        if self.settings.os == "Linux":
+            self.settings.compiler.libcxx = "libstdc++11"
```
Comment (ericLemanissier) [inline test_package/conanfile.py]: why is this necessary ?
Could we instead raise ConanInvalidConfiguration in the main recipe if libcxx is not supported ?

File: build.py [hunk 3]
```diff
@@ -1,12 +1,17 @@
...
-
+import os
+import platform
 from bincrafters import build_template_default
 
 if __name__ == "__main__":
 
-    builder = build_template_default.get_builder()
+    if platform.system() == "Linux":
+        command = os.environ["CONAN_DOCKER_ENTRY_SCRIPT"] + "; sudo apt-get -qq update && sudo apt-get -qq install -y python-minimal"
```
Comment (ericLemanissier) [inline build.py:11]: Why is `python-minimal` needed ? It should already be installed as a conan requirement right ? Also, `python-minimal` seems to install python 2 on some distros, which end-of-life is passed

Comment (balintfodor) [reply build.py:11]: Yes, the first problem I encountered is the missing `python2` dependency. https://github.com/bincrafters/conan-crashpad/runs/1746949242#step:5:654 on `testing/20200508`.

As I've learned, the chromium project's `luci/vpython` tool lets the scripts define their own virtual environment (much like pipenv or venv) to support scripts from the past. Looks like the tool `download_from_google_storage` is one that wants to run on py2 but the docker images that try to build the project don't include `python2` which is understandable. Other than installing manually in `build.py` I had no other idea to get it fixed.

File: conanfile.py [hunk 4]
```diff
@@ -43,8 +43,8 @@ def _make_spec(self):
 
     def configure(self):
         # It's not a C project, but libcxx is hardcoded in the project
-        del self.settings.compiler.libcxx
-        del self.settings.compiler.cppstd
+        # del self.settings.compiler.libcxx
+        # del self.settings.compiler.cppstd
```
Comment (ericLemanissier) [inline conanfile.py]:
```suggestion
        # del self.settings.compiler.cppstd
        pass
```
this shouldfix the IndentationError

File: test_package/conanfile.py (hunk 2 above)
Comment (balintfodor) [reply test_package/conanfile.py]: I'm actually very not sure about these two lines. If I don't specify `libstdc++11` here then the test package won't link with `undefined reference to` std stuff that is usually libstd incompatibility. I assume that `conan_basic_setup` in the test's `CMakeFile.txt` does something that is not done in the `.gn` build files of `crashpad` and `mini_chromium`.

I couldn't find the right way to tell the test package to use `libstdc++11`.

In the main recipe, there's an explicit `del self.settings.compiler.libcxx` in the `configure` function that I don't get either.

So you're suggesting to throw in `configure` like

```py
if self.settings.os == "Linux" and self.settings.compiler.libcxx != "libstdc++11":
   raise ConanInvalidConfiguration()
```
?

Comment (ericLemanissier) [reply test_package/conanfile.py]: to be short: yes !

File: build.py [hunk 5]
```diff
@@ -1,12 +1,17 @@
...
+        builder = build_template_default.get_builder(docker_entry_script=command)
```
Comment (ericLemanissier) [inline build.py]:
```suggestion
        builder = build_template_default.get_builder(docker_entry_script=command, pure_c=False)
```

File: build.py [hunk 6]
```diff
@@ -1,12 +1,17 @@
...
+    else:
+        builder = build_template_default.get_builder()
```
Comment (ericLemanissier) [inline build.py]:
```suggestion
        builder = build_template_default.get_builder(pure_c=False)
```

**Review by ericLemanissier — CHANGES_REQUESTED** [review]
The build passed because no configuration were built. With this change, it should try to build different libstdc++ settings

**Comment by ericLemanissier** [comment]
I'm wondering if we should use https://docs.conan.io/en/latest/reference/generators/compiler_args.html so that conan generates all the compiler arguments for the current settings, and then pass these to gn

**Comment by ericLemanissier** [comment]
@balintfodor I think the problem originates from the fact that gn does not conform to conan settings, because there is nothing in place to transmit the information from conan to gn. The test recipe is probably ok.

**Comment by balintfodor** [comment]
I'm definitely not an expert on `gn` but it looks to me that passing compiler parameters to `gn gen` would involve editing the `BUILD.gn` files which I'd like to avoid now. That needs a firm grasp of the whole build architecture of the crashpad project and the mini_chromium one.
There might be an easier solution but I don't see how to pass the conan generated compiler parameters directly to `gn`.

**Comment by ericLemanissier** [comment]
you should be able to pass additional flags via arguments: https://github.com/bincrafters/conan-crashpad/blob/testing/20200508/conanfile.py#L96-L100. Conan is able to generate compilation flags with https://docs.conan.io/en/latest/reference/generators/compiler_args.html

File: conanfile.py [hunk 7]
```diff
@@ -98,6 +98,13 @@ def _setup_args_gn(self):
         self._set_env_arg(args, "CXXFLAGS", "extra_cflags_cc")
         self._set_env_arg(args, "CXXFLAGS", "extra_cflags_objcc")
         self._set_env_arg(args, "LDFLAGS",  "extra_ldflags")
+        
+        args += [ "custom_conan_compiler_args_file=\\\"%s\\\"" % "@conanbuildinfo.args" ]
```
Comment (ericLemanissier) [inline conanfile.py]:
```suggestion
        args += [ "custom_conan_compiler_args_file=\\\"@%s\\\"" % os.path.join(self.install_folder, "conanbuildinfo.args") ]
```
With an absolute path, the file should always be found

**Comment by balintfodor** [comment]
Cool, GCC8,9 and Clang 6-9 are green again with the new changes. I'll clean up the commits and see what happened to the Windows builds.

It's rather strange, there's no reference to `depot_tools/cci.20201009` (it's still (depot_tools_installer/20200515@bincrafters/stable)) on this branch, but https://ci.appveyor.com/project/bincrafters/conan-crashpad/builds/38305755/job/lkyf8r9iwv1pcdb2#L254 tries to fetch that.

**Comment by balintfodor** [comment]
I've rebased on the current `testing/20200508` branch as https://github.com/bincrafters/conan-crashpad/commit/614aef32adddc2eabc7a3b6c6a3c888a86bb5f3f got in since I branched. Turns out it broke the Windows builds and it's something to do with `depot_tools/cci.20201009`. I guess that's another rabbit hole so I leave that for another PR.

@ericLemanissier, thanks for the help and hints! I think this PR is getting closer to be merged.

**Comment by balintfodor** [comment]
BTW I can't see Travis CI in the PR builder list here, that supposed to check the macOS configs if I'm right. It's ok for my branch https://travis-ci.org/github/balintfodor/conan-crashpad/builds/763552386 but is it some misconfiguration for this repo/PR?

**Comment by ericLemanissier** [comment]
yes, the macos builds will happen on github action if you delete the .travis.yml file (I just did it on the `testing/20200508` branch)

**Comment by balintfodor** [comment]
> yes, the macos builds will happen on github action if you delete the .travis.yml file (I just did it on the `testing/20200508` branch)

Thanks, I rebased on current `testing/20200508` then. macOS Apple Clang 10, 11 got into the queue indeed. Let's see what happens.

**Comment by balintfodor** [comment]
GHA approved the mac builds too 🥳 . @ericLemanissier, do we need other reviewers or is there anything else I need to do?

File: conanfile.py [hunk 8]
```diff
@@ -17,14 +17,15 @@ class CrashpadConan(ConanFile):
...
+    generators = "compiler_args"
 
     _commit_id = "c7d1d2a1dd7cf2442cbb8aa8da7348fa01d54182"
     _source_dir = "crashpad"
     _build_name = "out/Conan"
     _build_dir = os.path.join(_source_dir, _build_name)
 
     def build_requirements(self):
-        self.build_requires("depot_tools/cci.20201009")
+        self.build_requires("depot_tools_installer/20200515@bincrafters/stable")
```
Comment (ericLemanissier) [inline conanfile.py:35]: Can you please remind me why this change ?

Comment (balintfodor) [reply conanfile.py:35]: https://github.com/bincrafters/conan-crashpad/pull/13#issuecomment-802817025 and it's also in the commit message of https://github.com/bincrafters/conan-crashpad/pull/13/commits/74fb528e0237ad83c076a9a7999b408c251303b9.

Comment (ericLemanissier) [reply conanfile.py:35]: this looks like https://bugs.chromium.org/p/chromium/issues/detail?id=1067590#c7
It's probably worth reporting the issue on https://github.com/conan-io/conan-center-index/

**Comment by ericLemanissier** [comment]
let's summon other contributors for review. What do you think @reneme @Trafo @Tymolc @uilianries @Croydon  ?

EDIT: the CI failure on GCC < 8 and clang < 6.0 happens because the ubuntu versions used have glibc < 2.27 which does not support memfd_create

**Comment by Croydon** [comment]
The [reminder bot does not seem to work](https://github.com/probot/reminders/issues/51). Anyway, please ping me next month again if I don't remember it myself

**Comment by ericLemanissier** [comment]
@Croydon repinging you, as you asked in your last message

**Comment by Croydon** [comment]
Thanks for the ping. So a lot of issues you discussed seem to be solved in the meantime.

Is Crashpad still supporting GCC 4.9? It seems to be the only target not working right now

**Comment by ericLemanissier** [comment]
the error is `error: body of constexpr function 'static constexpr crashpad::Annotation::Type crashpad::Annotation::UserDefinedType(uint16_t)' not a return-statement`, so it probably requires c++14 ?

EDIT: it should already be using c++14: https://github.com/bincrafters/conan-crashpad/pull/13/files#diff-8f36dbc213e8b7050ce5f22d51d4e1c6d4e1be1d31f95f7ed25c7096ec4a581dR109, I guess we should simply ignore gcc < 5. a lot of packages do this already anyway

**Comment by balintfodor** [comment]
@Croydon, @ericLemanissier, sorry that it took so long. I've switched off `gcc` < 5 as you've suggested and I've also introduced a new option that can be used to force building with the `zlib` that crashpad comes with. Configuring `zlib` as a conan dependency would be the best way but for now I've found this low-effort one useful for my project. Maybe in some next PR.

Please let me know if there's anything you need and what's the next step to get this version with `crashpad/2021XXXX@bincrafters/stable` in my project :)

File: conanfile.py [hunk 9]
```diff
@@ -52,10 +56,15 @@ def source(self):
...
-        patch_base = os.path.join(self._source_dir, "third_party/mini_chromium/mini_chromium");
+        patch_base = os.path.join(
+            self._source_dir, "third_party/mini_chromium/mini_chromium")
         tools.patch(base_path=patch_base,
                     patch_file="patches/buildsystem-adaptions.patch")
 
+        if self.options.force_embedded_zlib:
+            tools.patch(base_path=patch_base,
+                        patch_file="patches/force-embedded-zlib.patch")
+
```
Comment (ericLemanissier) [inline conanfile.py]: this does not work in the general case, because conan reuses the same sources, regardless of the options used. In short the options values must never be used in `def source(self):`. This patching should occur in `def build(self)` instead

**Comment by balintfodor** [comment]
I promise I don't intend to do any other code changes 😃  @ericLemanissier @Croydon I guess ✅ ?

**Comment by ericLemanissier** [comment]
It looks fine overall, but two issues make me itchy:

- [the impossibility to use CCI's depot tools](https://github.com/bincrafters/conan-crashpad/pull/13#discussion_r598654897)
- [the recipe sometimes uses embedded zlib instead of conan provided zlib](https://github.com/bincrafters/conan-crashpad/pull/13/commits/ebaa1ea078db46c8d7517a962457c4a36d8a1b75)

WDYT @Croydon ?

File: conanfile.py [hunk 10]
```diff
@@ -15,8 +15,12 @@ class CrashpadConan(ConanFile):
...
     topics = ("conan", "crash-reporting", "logging", "minidump", "crash")
     settings = "os", "compiler", "build_type", "arch"
-    options = {'linktime_optimization': [True, False]}
-    default_options = {"linktime_optimization": False}
+    options = {
+        "linktime_optimization": [True, False],
+        "force_embedded_zlib": [True, False]}
+    default_options = {
+        "linktime_optimization": False,
+        "force_embedded_zlib": False}
```
Comment (Croydon) [inline conanfile.py:23]: What is the advantage of using the embedded zlib?

Comment (balintfodor) [reply conanfile.py:23]: Crashpad's zlib is configured (https://chromium.googlesource.com/crashpad/crashpad/+/refs/heads/master/third_party/zlib/BUILD.gn#17) to switch the zlib source based on the platform you're building on. That's not very transparent dependency management. In my project, I want to have a grip on where the libs are coming from and I prefer not pulling the system's current zlib version (see `zlib_source = "system"` in the link).

So I've put this option here. It doesn't do any harm, setting it `false` does exactly the same thing that before but `true` is a huge help when one needs to make sure where zlib comes from.

As I said, the best would be making crashpad's zlib to be a proper conan dependency but that's out of scope for me in this PR, it's already fixing a lot more than I first intended (which is good).

Comment (Croydon) [reply conanfile.py:23]: Fine by me for now, though, in the long term, every single patch more is a burden 🙂

**Review by Croydon — APPROVED** [review]
Good enough for now. If @ericLemanissier is also okay with it we can merge it here.

Please note, however, that we are currently in some ongoing infrastructure changes. The upload of the packages after the merge in this git repository will partially fail. I will move this recipe to [our central repository](https://github.com/bincrafters/community) afterwards. It will be available from [our new Conan remote](https://bincrafters.readthedocs.io/en/latest/using_packages.html#adding-the-bincrafters-repository-as-a-conan-remote). There is an upcoming blog post explaining those changes: https://github.com/bincrafters/bincrafters.github.io/pull/19

**Review by ericLemanissier — APPROVED** [review]
Thanks for the great work !

**Comment by Croydon** [comment]
Migration pull request: https://github.com/bincrafters/community/pull/1381
